src/jumps/jumpTracker.ts

```diff
@@ -0,0 +1,265 @@
+   */
+  public get currentJumpNumber(): number {
+    return this._currentJumpNumber;
+  }
+
+  /**
+   * Current jump in the list of jumps.
+   */
+  public get currentJump(): Jump {
+    return this._jumps[this._currentJumpNumber] || null;
+  }
+
+  /**
+   * Current jump in the list of jumps.
+   */
+  public get hasJumps(): boolean {
+    return this._jumps.length > 0;
+  }
+
+  /**
+   * Last jump in list of jumps.
+   */
+  public get end(): Jump | null {
+    return this._jumps[this._jumps.length - 1];
+  }
+
+  /**
+   * First jump in list of jumps.
+   */
+  public get start(): Jump | null {
+    return this._jumps[0];
+  }
+
+  /**
+   * Record that a jump occurred.
+   *
+   * If the current position is back in history,
+   * jumps after this position will be removed.
+   *
+   * @param from - File/position jumped from
+   * @param to - File/position jumped to
+   */
+  public recordJump(from: Jump | null, to?: Jump | null) {
+    if (from && to && from.onSameLine(to)) {
+      return;
+    }
+
+    if (from) {
+      this.clearJumpsOnSameLine(from);
+    }
+
+    if (from && !from.onSameLine(to)) {
+      this._jumps.push(from);
+    }
+
+    this._currentJumpNumber = this._jumps.length;
+
+    this.clearOldJumps();
+  }
+
+  /**
+   * Record that a jump occurred from one file to another.
+   * This is likely only needed on a handler for
+   * vscode.window.onDidChangeActiveTextEditor.
+   *
+   * File jumps have extra checks in place, keeping in mind
+   * whether this plugin initiated the jump, whether the new file is
+   * a legitimate file.
+   *
+   * @param from - File/position jumped from
+   * @param to - File/position jumped to
+   */
+  public recordFileJump(from: Jump | null, to: Jump) {
+    if (this.isJumpingFiles) {
+      this.isJumpingFiles = false;
+      return;
+    }
+
+    if (to.editor && to.editor.document && to.editor.document.isClosed) {
+      // Wallaby.js seemed to be adding an extra file jump, named e.g. extension-output-#4
+      // It was marked closed when jumping to it. Hopefully we can rely on checking isClosed
+      // when extensions get all weird on us.
+      return;
+    }
+
+    if (from && from.fileName === to.fileName) {
+      return;
+    }
+
+    if (from) {
+      this.clearJumpsOnSameLine(from);
+    }
+
+    if (from && !from.onSameLine(to)) {
+      this._jumps.push(from);
+    }
+
+    this._currentJumpNumber = this._jumps.length;
+
+    this.clearOldJumps();
+  }
+
+  /**
+   * Get the previous jump in history.
+   * Continues further back if the current line is on the same line.
+   *
+   * @param from - File/position jumped from
+   */
+  public jumpBack(from: Jump): Jump {
+    if (!this.hasJumps) {
+      return from;
+    }
+
+    if (this._currentJumpNumber <= 0) {
+      return this._jumps[0];
+    }
+
+    const to: Jump = this._jumps[this._currentJumpNumber - 1];
+
+    if (this._currentJumpNumber === this._jumps.length) {
+      this.recordJump(from, to);
```

Comment:
Why is this necessary?

Reply:
That is how vanilla Vim appears to work, which lets you `<C-i>` after you've jumped back. Vim doesn't keep track of where you jumped to until you jump back. 

Here's video demonstrating that in Vim: https://cl.ly/b1d14cd87b45

Notice how the first time I show `:jump`, it doesn't show the current line of where I jumped to (`export class JumpTracker`), and that the `>` position arrow is outside of the jumps list, but then I press `<C-o>` and I can see a jump was added and the `>` position arrow is now on the `from` jump.